Add event tracking to projects

jsvgoncalves opened the pull request:
This PR adds event sourcing to brightway2-data.

It does this by:
1. creating a parent ORM class, `SignaledDataset`, that other ORM classes can inherit from.
2. adding signals for events which can be used to source the database
3. adding a `revision`s to each project, that use snowflake ids for distributed id generation


```
{
"metadata": {
  "revision": <this-revision-id>
  "parent_revision": <parent-revision-id>
  "title": "<optional>"
  "description": "<optional>"
  "authors": "<optional>" (maybe shouldn't be optional)
},
"data": […]
}
```

jsvgoncalves committed:
Initialise event tracking to projects

* Datasets now inherit from a `SignaledDataset` that generates revisions
  on `.save()`, and sends a signal.
* Projects now have two extra fields `is_sourced` to enable automatic
  revisioning, and `revision` to keep track of revisions.
* Loading revisions will come at a later commit.

Signed-off-by: João Gonçalves <jsvgoncalves@gmail.com>

jsvgoncalves commented on bw2data/project.py:
```diff
@@ -52,13 +56,52 @@ def __lt__(self, other):
+    def dir(self):
+        return Path(self._get_base_directories()) / safe_filename(
+            self.name, full=self.full_hash
+        )
```
This is copied over from `ProjectManager`, we need to decide if the patches go on DB or on a file, then I'll refactor. However, I think that projects should be more encapsulated and be responsible for their own internal management, and `ProjectManager` can play a more facade-ish role.

jsvgoncalves replied:
@bbguimaraes

jsvgoncalves commented on bw2data/project.py:
```diff
@@ -52,13 +56,52 @@ def __lt__(self, other):
+        self.revision = uuid.uuid4()
+        self.save()
+
+    def add_revision(self, patch: str, revision: str | None = None) -> None:
```
```suggestion
    def add_revision(self, patch: DeepDiff, revision: str | None = None) -> None:
```

This would be another possibility, we have to decide who has the responsibility of applying future patches. If it's the `Project`, then there's a good chance it should know about `DeepDiff` instead just `str` to files…

jsvgoncalves reviewed (commented):
@cmutel see below

cmutel commented on bw2data/project.py:
```diff
@@ -35,6 +39,10 @@ def lockable():
 class ProjectDataset(Model):
+    # Event sourcing
+    is_sourced = BooleanField(default=False, constraints=[SQL("DEFAULT 0")])
+    revision = TextField(null=True, default=uuid.uuid4)
```
Maybe instead:
```python
def uuid4_func() -> str:
    return uuid.uuid4().hex
    
revision = TextField(null=True, default=uuid4_func)
```

I guess this would work but the implicit type conversion feels icky.

jsvgoncalves replied:
stale, we use snowflake id

cmutel commented on bw2data/project.py:
```diff
@@ -35,6 +39,10 @@ def lockable():
 
 class ProjectDataset(Model):
+    # Event sourcing
+    is_sourced = BooleanField(default=False, constraints=[SQL("DEFAULT 0")])
```
[Peewee schema migrations](https://docs.peewee-orm.com/en/latest/peewee/playhouse.html#schema-migrations).

jsvgoncalves replied:
Cool, I'll add this in one of the next commits, cheers.

cmutel commented on bw2data/project.py:
```diff
@@ -54,13 +62,83 @@ def __lt__(self, other):
         else:
             return self.name.lower() < other.name.lower()
 
+    def _get_base_directories(self):
```
I don't quite understand how to read the diff here - the method `_get_base_directories` should already exist, but I don't see any lines removed? What am I missing?

jsvgoncalves replied:
https://github.com/brightway-lca/brightway2-data/pull/174#discussion_r1644553443 see this earlier comment

cmutel commented on bw2data/signals.py:
```diff
@@ -1,5 +1,9 @@
 from blinker import signal
 
+database_saved = signal(
```
`database_save` - is this master-level trolling? 😛

jsvgoncalves replied:
There are two hard things in computer science… No but seriously, this name will change later.

bbguimaraes committed:
Add dataset attribute to proxy classes

Co-authored-by: Chris Mutel <cmutel@gmail.com>

bbguimaraes committed:
Add method to apply a database revision

bbguimaraes committed:
Add helper class to traverse revision graphs

bbguimaraes committed:
Extract revision patch generation function

cmutel committed:
Ensure compatibility with older python versions

* Add typing extensions for `@override` and `Self`
* Remove unnecessary `(object)`

bbguimaraes committed:
Extract JSON encoder for revisions

bbguimaraes committed:
Extract delta creation function

bbguimaraes committed:
Add type for revision delta

bbguimaraes committed:
Move object type and ID from metadata to delta

bbguimaraes committed:
Extract function to create revision

bbguimaraes committed:
Include multiple delta in a single revision

jsvgoncalves committed:
Use snowflake ID for revisions

* This theoretically allows IDs to be sortable.

Signed-off-by: João Gonçalves <jsvgoncalves@gmail.com>

cmutel committed:
Add `ProjectDataset` database migrations

* Add migration for new fields on `ProjectDataset`
* Refactor previous full_hash migration
* Add tests for migrations

jsvgoncalves committed:
Allow first revision to be orphan

Signed-off-by: João Gonçalves <jsvgoncalves@gmail.com>

jsvgoncalves committed:
Test everything under the tests directory

Signed-off-by: João Gonçalves <jsvgoncalves@gmail.com>

jsvgoncalves committed:
Add load_revisions to sync changes from dir

* Add an optional boolean to `SignaledDataset.save()` to deactivate the
  signal emission.
* Change `revision_id` type from `str` to `int` everywhere
* Write a `revisions/head` file containing the id of the current
  revision, so it can be used for less inefficient graph traversal.

Signed-off-by: João Gonçalves <jsvgoncalves@gmail.com>

bbguimaraes committed:
Add tests for `load_revisions`

cmutel committed:
Fix error if no revisions present

bbguimaraes committed:
Fix type hints

bbguimaraes committed:
Add an `__init__` method to `Delta`

This helps with type annotations, since Python does not have a syntax for
multiple constructor methods.  It also has no good way to signal that `__init__`
should not be called directly, which makes this less than ideal.

bbguimaraes committed:
Accept initial metadata when creating revisions

bbguimaraes committed:
Extract delta generation from `add_revision`

This allows (multiple) delta to be aggregated and included in a single revision.

cmutel committed:
Add more event types (#204)

* Use different signals for different actions
* Add activity code change event
* Allow more flexible revisions dict dumping
* Add event sourcing support for copying nodes

Co-authored-by: Chris Mutel <cmutel@gmail.com>

jsvgoncalves commented:
@cmutel @bbguimaraes have fun with the 2k lines, kthxbye.

No seriously, please just have (yet) another look, and I suggest we merge this (`no-ff` please) so that we can then also merge in #202. Any issues found during more extensive testing can be handled in their own PRs.

cmutel reviewed (approved):
LGTM. Still some polishing and this isn't complete but merging and then fixing up later is fine.

I checked through the code and didn't find anything blocking.
